File: mobile/test/order_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:restaurant_mobile/core/auth/auth_api.dart';
import 'package:restaurant_mobile/core/orders/order.dart';
import 'package:restaurant_mobile/core/orders/order_api.dart';

String loiJson(String code) => jsonEncode({
      'error': {'code': code, 'message': 'in English', 'details': {}}
    });

HttpOrderApi apiTraVe(int status, String body,
        {void Function(http.Request)? ghiLai}) =>
    HttpOrderApi(
      baseUrl: 'http://test',
      client: MockClient((request) async {
        ghiLai?.call(request);
        return http.Response(body, status,
            headers: {'content-type': 'application/json; charset=utf-8'});
      }),
    );

void main() {
  group('nhãn trạng thái đơn', () {
    test('Ready nói rõ là CHỜ MANG RA, không phải đã xong bữa', () {
      // Dịch Ready thành "Hoàn tất" sẽ khiến khách tưởng có thể đứng dậy đi về trong khi món
      // còn nằm ở bếp.
      expect(nhanTrangThaiDon('Ready'), 'Nấu xong, chờ mang ra');
      expect(nhanTrangThaiDon('Ready'), isNot(contains('Hoàn tất')));
    });

    test('Completed mới là đã thanh toán', () {
      expect(nhanTrangThaiDon('Completed'), 'Đã thanh toán');
    });

    test('phủ hết trạng thái backend có', () {
      // Thiếu một trạng thái nghĩa là khách thấy chữ tiếng Anh giữa màn hình tiếng Việt.
      for (final s in [
        'Draft',
        'Placed',
        'Confirmed',
        'Preparing',
        'Ready',
        'Served',
        'Completed',
        'Cancelled'
      ]) {
        expect(nhanTrangThaiDon(s), isNot(equals(s)),
            reason: 'thiếu nhãn cho $s');
      }
    });

    test('trạng thái LẠ trả nguyên văn, không nuốt thành câu chung chung', () {
      // Backend có thể thêm trạng thái mới trước khi app kịp cập nhật. Hiện "Đang xử lý" cho mọi
      // thứ chưa biết sẽ giấu mất chuyện đó và không ai phát hiện app đã lạc hậu.
      expect(nhanTrangThaiDon('TrangThaiMoi'), 'TrangThaiMoi');
    });
  });

  group('nhãn trạng thái món', () {
    test('Pending ở cấp MÓN là chờ nấu, không phải chờ thu tiền', () {
      // Cùng chữ Pending, hai nghĩa khác nhau ở hai cấp. Dùng chung một câu là cách nhanh nhất
      // để hiểu nhầm.
      expect(nhanTrangThaiMon('Pending'), 'Chờ nấu');
    });

    test('phủ hết trạng thái món', () {
      for (final s in [
        'Pending',
        'Preparing',
        'Ready',
        'Served',
        'Cancelled'
      ]) {
        expect(nhanTrangThaiMon(s), isNot(equals(s)),
            reason: 'thiếu nhãn cho $s');
      }
    });
  });

  group('đơn đã xong chưa', () {
    test('Completed và Cancelled là xong', () {
      expect(donDaXong('Completed'), isTrue);
      expect(donDaXong('Cancelled'), isTrue);
    });

    test('Served CHƯA xong — món đã ra bàn nhưng chưa trả tiền', () {
      // Đây là chỗ dễ sai nhất: "đã mang ra bàn" nghe như kết thúc, nhưng hoá đơn vẫn mở.
      expect(donDaXong('Served'), isFalse);
    });
  });

  group('gọi API', () {
    test('uỷ quyền bằng X-Table-Session-Token, KHÔNG bằng JWT', () async {
      // Đơn thuộc về cái BÀN, không thuộc về tài khoản. Ai đang ngồi ở bàn đều xem được, kể cả
      // khách vãng lai đi cùng — đúng như web.
      http.Request? daGui;
      await apiTraVe(200, '{"orders":[],"total":0}', ghiLai: (r) => daGui = r)
          .donCuaPhien('ts_abc', 'tst_bi_mat');

      expect(daGui!.url.path, '/api/table-sessions/ts_abc/orders');
      expect(daGui!.headers['X-Table-Session-Token'], 'tst_bi_mat');
      expect(daGui!.headers.containsKey('Authorization'), isFalse);
    });

    test('phân giải đơn và món', () async {
      final ds = await apiTraVe(
          200,
          jsonEncode({
            'orders': [
              {
                'orderId': 'o1',
                'orderCode': 'DH0001',
                'status': 'Preparing',
                'totalAmount': 175000,
                'createdAt': '2026-08-20T12:00:00.123456789Z',
                'items': [
                  {
                    'orderItemId': 'oi1',
                    'name': 'Phở bò tái',
                    'quantity': 2,
                    'unitPrice': 60000,
                    'lineTotal': 120000,
                    'status': 'Preparing',
                  }
                ],
              }
            ],
            'total': 1,
          })).donCuaPhien('ts_abc', 'tst');

      expect(ds.single.orderCode, 'DH0001');
      expect(ds.single.createdAt.isUtc, isTrue);
      expect(ds.single.items.single.name, 'Phở bò tái');
      expect(ds.single.items.single.quantity, 2);
    });

    test('phiên HẾT HẠN có câu riêng, khác hẳn token sai', () async {
      // Khách không làm gì sai, chỉ là bàn đã đóng. Báo "token không hợp lệ" khiến họ tưởng app
      // hỏng và thử đi thử lại.
      try {
        await apiTraVe(410, loiJson('TABLE_SESSION_EXPIRED'))
            .donCuaPhien('ts', 'tst');
        fail('phải ném lỗi');
      } on AuthException catch (e) {
        expect(e.code, 'TABLE_SESSION_EXPIRED');
        expect(e.message, contains('đã kết thúc'));
      }
    });

    test('token sai cho câu bảo quét lại QR', () async {
      await expectLater(
        apiTraVe(401, loiJson('TABLE_SESSION_TOKEN_INVALID'))
            .donCuaPhien('ts', 'sai'),
        throwsA(isA<AuthException>()
            .having((e) => e.code, 'code', 'TABLE_SESSION_TOKEN_INVALID')
            .having((e) => e.message, 'message', contains('Quét lại'))),
      );
    });

    test('phiên chưa có đơn nào là hợp lệ, không phải lỗi', () async {
      expect(
          await apiTraVe(200, '{"orders":[],"total":0}')
              .donCuaPhien('ts', 'tst'),
          isEmpty);
    });
  });
}

File: mobile-rn/src/core/orders/__tests__/order.test.ts
import {
  chophepHuyMon,
  donDaXong,
  moTaBepDong,
  moTaUocLuong,
  nhanTrangThaiDon,
  nhanTrangThaiMon,
} from '../order';

describe('nhãn trạng thái đơn', () => {
  it('Ready nói rõ là CHỜ MANG RA, không phải đã xong bữa', () => {
    // Dịch thành "Hoàn tất" sẽ khiến khách tưởng có thể đứng dậy đi về.
    expect(nhanTrangThaiDon('Ready')).toBe('Nấu xong, chờ mang ra');
    expect(nhanTrangThaiDon('Ready')).not.toContain('Hoàn tất');
  });

  it('Completed mới là đã thanh toán', () => {
    expect(nhanTrangThaiDon('Completed')).toBe('Đã thanh toán');
  });

  it('phủ hết trạng thái backend có', () => {
    for (const s of [
      'Draft',
      'Placed',
      'Confirmed',
      'Preparing',
      'Ready',
      'Served',
      'Completed',
      'Cancelled',
    ]) {
      expect(nhanTrangThaiDon(s)).not.toBe(s);
    }
  });

  it('trạng thái LẠ trả nguyên văn, không nuốt thành câu chung chung', () => {
    // Backend có thể thêm trạng thái mới trước khi app kịp cập nhật. Hiện "Đang xử lý" cho mọi
    // thứ chưa biết sẽ giấu mất chuyện đó và không ai phát hiện app đã lạc hậu.
    expect(nhanTrangThaiDon('TrangThaiMoiCuaBackend')).toBe('TrangThaiMoiCuaBackend');
  });
});

describe('nhãn trạng thái món', () => {
  it('Pending ở cấp MÓN là chờ nấu, không phải chờ thu tiền', () => {
    expect(nhanTrangThaiMon('Pending')).toBe('Chờ nấu');
  });

  it('phủ hết trạng thái món', () => {
    for (const s of ['Pending', 'Preparing', 'Ready', 'Served', 'Cancelled']) {
      expect(nhanTrangThaiMon(s)).not.toBe(s);
    }
  });

  it('trạng thái món lạ cũng trả nguyên văn', () => {
    expect(nhanTrangThaiMon('MonLa')).toBe('MonLa');
  });
});

describe('đơn đã xong chưa', () => {
  it('Completed và Cancelled là xong', () => {
    expect(donDaXong('Completed')).toBe(true);
    expect(donDaXong('Cancelled')).toBe(true);
  });

  it('Served CHƯA xong — món đã ra bàn nhưng chưa trả tiền', () => {
    expect(donDaXong('Served')).toBe(false);
    expect(donDaXong('Ready')).toBe(false);
  });
});

describe('ước lượng thời gian (hạn chế #10)', () => {
  it('KHÔNG có ước lượng thì trả null — app không được bịa con số', () => {
    // Ba điều kiện của #10 (không đoán bừa, luôn là khoảng, có tính tải bếp) vẫn giữ nguyên qua
    // lần viết lại ở #141. Một con số bịa ở tầng app phá cả ba mà không ai thấy.
    expect(moTaUocLuong(null, null)).toBeNull();
    expect(moTaUocLuong(10, null)).toBeNull();
    expect(moTaUocLuong(null, 20)).toBeNull();
  });

  it('có ước lượng thì hiện dạng KHOẢNG, không phải một con số', () => {
    expect(moTaUocLuong(15, 25)).toBe('15–25 phút');
  });

  it('khoảng suy biến (low >= high) vẫn nói rõ là "khoảng"', () => {
    // Một con số trần trụi hứa nhiều hơn thứ hệ thống biết.
    expect(moTaUocLuong(12, 12)).toBe('khoảng 12 phút');
    expect(moTaUocLuong(12, 10)).toBe('khoảng 12 phút');
  });
});

describe('báo bếp đang đông', () => {
  it('bếp bình thường thì KHÔNG nói gì', () => {
    expect(moTaBepDong(false, '15–25 phút')).toBeNull();
  });

  it('bếp đông thì nói RÕ VÌ SAO lâu', () => {
    // Ước lượng nhảy từ 15–25 lên 42–57 phút mà không nói vì sao trông như app tính sai.
    expect(moTaBepDong(true, '42–57 phút')).toContain('Bếp đang đông');
  });

  it('KHÔNG báo bếp đông khi chưa có ước lượng', () => {
    // Báo "bếp đang đông" mà không kèm con số nào là gieo lo lắng mà không cho khách thứ gì để
    // quyết định.
    expect(moTaBepDong(true, null)).toBeNull();
  });
});

describe('huỷ món (hạn chế #11)', () => {
  it('CHỈ huỷ được món đang Pending', () => {
    // Nhân viên vẫn huỷ được món Preparing, khách thì không — tới lúc đó bếp đã dùng nguyên liệu.
    expect(chophepHuyMon('Pending', true)).toBe(true);
    expect(chophepHuyMon('Preparing', true)).toBe(false);
    expect(chophepHuyMon('Ready', true)).toBe(false);
    expect(chophepHuyMon('Served', true)).toBe(false);
    expect(chophepHuyMon('Cancelled', true)).toBe(false);
  });

  it('KHÔNG có token của đơn thì không huỷ được, dù món đang Pending', () => {
    // Đơn do máy khác trong bàn đặt thì máy này không có token — và đó là đúng: người đặt mới là
    // người quyết định huỷ.
    expect(chophepHuyMon('Pending', false)).toBe(false);
  });
});
